### PR TITLE
add decision transformer

### PR DESCRIPTION
**Pull request by TomorrowIsAnOtherDay**

### CODE REVIEW

File: benchmark/torch/DT/train.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import random
+import sys
+
+from evaluate_episodes import eval_episodes
+from tqdm import tqdm
+import parl
+from model import TrajectoryModel
+from agent import DTAgent
+from parl.utils import summary
+from evaluate_episodes import eval_episodes
```
Comment (wangzelong0663) [inline benchmark/torch/DT/train.py:28]: import repeated  same to row 23.

File: benchmark/torch/DT/train.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        n_head=config['n_head'],
+        n_inner=4 * config['embed_dim'],
+        activation_function=config['activation_function'],
+        n_positions=1024,
+        resid_pdrop=config['dropout'],
+        attn_pdrop=config['dropout'],
+    )
+
+    model = model.to(device=device)
+    alg = parl.algorithms.DecisionTransformer(model, config['learning_rate'], config['warmup_steps'],
```
Comment (wangzelong0663) [inline benchmark/torch/DT/train.py:75]: replaced by `from parl.algorithms import DecisionTransformer` may be better.

File: benchmark/torch/DT/data_loader.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        logger.info(f'Average return: {np.mean(returns):.2f}, std: {np.std(returns):.2f}')
+        logger.info(f'Max return: {np.max(returns):.2f}, min: {np.min(returns):.2f}')
+        self.trajectories = trajectories
+        self.num_trajectories = num_trajectories
+        self.p_sample = p_sample
+        self.sorted_inds = sorted_inds
+        self.state_dim = trajectories[0]['observations'].shape[1]
+        self.act_dim = trajectories[0]['actions'].shape[1]
+
+    def get_batch(self, batch_size=256, max_len=20):
```
Comment (wangzelong0663) [inline benchmark/torch/DT/data_loader.py:60]: put the `gamma=1` here as the params of the `get_batch` function maybe better.

File: benchmark/torch/DT/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            s.append(traj['observations'][si:si + max_len].reshape(1, -1, self.state_dim))
+            a.append(traj['actions'][si:si + max_len].reshape(1, -1, self.act_dim))
+            r.append(traj['rewards'][si:si + max_len].reshape(1, -1, 1))
+            if 'terminals' in traj:
+                d.append(traj['terminals'][si:si + max_len].reshape(1, -1))
+            else:
+                d.append(traj['dones'][si:si + max_len].reshape(1, -1))
+            timesteps.append(np.arange(si, si + s[-1].shape[1]).reshape(1, -1))
+            timesteps[-1][timesteps[-1] >= self.max_ep_len] = self.max_ep_len - 1  # padding cutoff
+            rtg.append(self.discount_cumsum(traj['rewards'][si:], gamma=1.)[:s[-1].shape[1] + 1].reshape(1, -1, 1))
```
Comment (wangzelong0663) [inline benchmark/torch/DT/data_loader.py:83]: As mentioned above, if changed above, then change `gamma=1` to `gamma=gamma`

File: benchmark/torch/DT/evaluate_episodes.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        action = agent.predict(
+            (states.to(dtype=torch.float32) - state_mean) / state_std,
+            actions.to(dtype=torch.float32),
+            rewards.to(dtype=torch.float32),
+            target_return.to(dtype=torch.float32),
+            timesteps.to(dtype=torch.long),
+        )
+        actions[-1] = action
+        action = action.detach().cpu().numpy()
```
Comment (wangzelong0663) [inline benchmark/torch/DT/evaluate_episodes.py]: maybe can put the tensor to numpy  in the `agent.py`.

File: benchmark/torch/DT/agent.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+class DTAgent(parl.Agent):
+    def __init__(self, algorithm, config):
+        super(DTAgent, self).__init__(algorithm)
+        self.dataset = None
+        self.config = config
+
+    def predict(self, states, actions, rewards, returns_to_go, timesteps, **kwargs):
+        action = self.alg.predict(states, actions, rewards, returns_to_go, timesteps)
+        return action
```
Comment (wangzelong0663) [inline benchmark/torch/DT/agent.py:31]: turn the tensor to cpu().numpy() here?

File: benchmark/torch/DT/agent.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    def learn(self):
+        batch_data = self.dataset.get_batch(self.config['batch_size'])
+        loss = self.alg.learn(*batch_data)
+        return loss
```
Comment (wangzelong0663) [inline benchmark/torch/DT/agent.py:37]: tensor to numpy here?

File: parl/algorithms/torch/dt.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        loss = torch.mean((action_preds - action_target)**2)
+
+        self.optimizer.zero_grad()
+        loss.backward()
+        torch.nn.utils.clip_grad_norm_(self.model.parameters(), .25)
+        self.optimizer.step()
+        self.scheduler.step()
+
+        return loss.detach().cpu().item()
```
Comment (wangzelong0663) [inline parl/algorithms/torch/dt.py]: put the `.detach().cpu().item()` in the agent.py?

File: benchmark/torch/DT/evaluate_episodes.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+        state, reward, done, _ = env.step(action)
+
+        cur_state = torch.from_numpy(state).to(device=device).reshape(1, state_dim)
+        states = torch.cat([states, cur_state], dim=0)
+        rewards[-1] = reward
+
+        pred_return = target_return[0, -1] - (reward / scale)
+        target_return = torch.cat([target_return, pred_return.reshape(1, 1)], dim=1)
+        timesteps = torch.cat([timesteps, torch.ones((1, 1), device=device, dtype=torch.long) * (t + 1)], dim=1)
```
Comment (wangzelong0663) [inline benchmark/torch/DT/evaluate_episodes.py:63]: put the numpy to tensor in the agent.py?

File: benchmark/torch/DT/evaluate_episodes.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        episode_return += reward
+        episode_length += 1
+
+        if done:
+            break
+
+    return episode_return, episode_length
+
+
+def eval_episodes(target_rew, env, state_dim, act_dim, agent, max_ep_len, scale, state_mean, state_std, device):
```
Comment (wangzelong0663) [inline benchmark/torch/DT/evaluate_episodes.py:74]: donn't need to set the random seeds in this  function?

File: benchmark/torch/DT/evaluate_episodes.py (hunk 5 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/evaluate_episodes.py]: done

File: benchmark/torch/DT/train.py (hunk 1 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/train.py:28]: removed.

File: benchmark/torch/DT/train.py (hunk 2 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/train.py:75]: I would like to show the users that it is perfect to import our algorithm within one line.

File: benchmark/torch/DT/data_loader.py (hunk 3 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/data_loader.py:60]: The code provided by the authors does not use decayed reward to compute `reward-to-go`. So it is ok not passing the `gamma` here.

File: benchmark/torch/DT/data_loader.py (hunk 4 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/data_loader.py:83]: The response to this comment can be found above.

File: benchmark/torch/DT/agent.py (hunk 6 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/agent.py:31]: done.

File: benchmark/torch/DT/agent.py (hunk 7 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/agent.py:37]: done

File: parl/algorithms/torch/dt.py (hunk 8 above)
Comment (TomorrowIsAnOtherDay) [reply parl/algorithms/torch/dt.py]: done.

File: benchmark/torch/DT/evaluate_episodes.py (hunk 10 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/DT/evaluate_episodes.py:74]: Yes. We run 100 episodes with different random seeds.  It is sufficient to evaluate the performance of an agent in a stochastic environment.

**Review by wangzelong0663 — APPROVED** [review]
